docs/developers/best_practices.md

```diff
@@ -0,0 +1,63 @@
+# IREE Best Practices
+
+This page contains a list of best practices for getting the most out of IREE,
+spanning model authoring, ahead-of-time compilation, and runtime use. Treat
+these as a collection of ideas to consider or areas to start benchmarking when
+working on your own applications.
+
+## Introduction
+
+Common themes include:
+
+* give the compiler as much information as possible
+* give the compiler opportunities to batch work together or defer computation
+* keep compute devices saturated with work through pipelining
+* use dense math where possible, particular for inner loop bodies
+* limit synchronization points between devices like CPUs and GPUs
+* profile early and often, using the right tools for each level of granularity
+
+## Practices for model authoring
+
+### Track state within your model when possible
+
+If your model is stateful, prefer to store that state directly within your
```

Comment:
grammar nit: remove all commas in this paragraph - they aren't needed with compound predicates